packages/substrate-client/src/chainhead/chainhead.ts

```diff
@@ -1,109 +1,158 @@
-import type { OperationEvents } from "./internal-types"
+import type { UnsubscribeFn } from "@/."
+import type {
+  ClientRequest,
+  ClientRequestCb,
+  FollowSubscriptionCb,
+} from "@/client"
+import type { OperationEvents, Stop } from "./internal-types"
 import type {
   ChainHead,
   FollowEventWithoutRuntime,
   FollowEventWithRuntime,
   FollowResponse,
 } from "./public-types"
+import {
+  Subscriber,
+  getSubscriptionsManager,
+  noop,
+  deferred,
+} from "@/internal-utils"
 import { createBodyFn } from "./body"
 import { createCallFn } from "./call"
 import { createHeaderFn } from "./header"
 import { createStorageFn } from "./storage"
 import { createUnpinFn } from "./unpin"
-import {
-  Subscriber,
-  getSubscriptionsManager,
-} from "@/utils/subscriptions-manager"
+import { ErrorDisjoint, ErrorStop } from "./errors"
 
-function isOperationEvent(
-  event: FollowEventWithRuntime | FollowEventWithoutRuntime | OperationEvents,
-): event is OperationEvents {
+type FollowEvent =
+  | FollowEventWithRuntime
+  | FollowEventWithoutRuntime
+  | OperationEvents
+  | Stop
+
+function isOperationEvent(event: FollowEvent): event is OperationEvents {
   return (event as OperationEvents).operationId !== undefined
 }
 
 export function getChainHead(
-  request: ClientRequest<
-    string,
-    FollowEventWithoutRuntime | FollowEventWithRuntime | OperationEvents
-  >,
+  request: ClientRequest<string, FollowEvent>,
 ): ChainHead {
   return (
     withRuntime: boolean,
-    cb:
+    onFollowEvent:
       | ((event: FollowEventWithoutRuntime) => void)
       | ((event: FollowEventWithRuntime) => void),
+    onFollowError: (e: Error) => void,
   ): FollowResponse => {
     const subscriptions = getSubscriptionsManager()
-    let unfollow: () => void = () => {}
-
-    let followSubscription: Promise<string> | string = new Promise((res) => {
-      unfollow = request(
-        "chainHead_unstable_follow",
-        [withRuntime],
-        (subscriptionId, follow) => {
-          const done = follow(subscriptionId, {
-            next: (event) => {
-              if (isOperationEvent(event)) {
-                subscriptions.next(event.operationId, event)
-              } else {
-                if (event.event === "stop") {
-                  done()
-                  subscriptions.errorAll(new Error("disjoint"))
-                }
-                cb(event as any)
-              }
-            },
-            error: (e) => {
-              subscriptions.errorAll(e)
-            },
-          })
-
-          unfollow = () => {
-            done()
-            subscriptions.errorAll(new Error("disjoint"))
-            request("chainHead_unstable_unfollow", [subscriptionId])
-          }
 
-          followSubscription = subscriptionId
-          res(subscriptionId)
-        },
-      )
-    })
+    const deferredFollow = deferred<string | Error>()
+    let followSubscription: Promise<string | Error> | string | null =
```

Comment:
i believe we discussed this offline but this is just a hack to do streams correct.

since it starts as an unresolved promise. the first time we resolve the promise it is the "first value" in the stream. But since we cannot resolve a promise twice, we simply reassign the variable for subsequent values.

Reply:
You're right in noting that this implementation is trying to approximate behavior typical of Observables.

The crux is not so much about being "unable to resolve a promise twice." We could retain a resolved Promise and keep leveraging it until the user opts to "unfollow" or an error occurs, subsequently setting its value to `null` to signify a dead subscription. 

However, following such an approach means any `chainHead` calls would **always** execute after a micro-task, rather than being evaluated synchronously whenever possible. Even though it sounds minor, such small deviations often lead to race conditions or unexpected consumer behaviors. Essentially, this would be a leaky abstraction. It inadvertently exposes users to the nitty-gritty of the implementation, complicating the library's usability.

Consider this scenario: If a user wishes to cancel an ongoing operation, the [`chainHead_unstable_stopOperation`](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_stopOperation.html#chainhead_unstable_stopoperation) should be invoked synchronously within that callback, not postponed to a later microtask. Always keep in mind, awaiting a resolved promise still results in a microtask, and it's never synchronous. This logic applies across all `chainHead` calls (like `unpin`, `header`, `call`, etc.). Our ultimate goal is to ensure synchronous logic execution when feasible.

Also, the variable `followSubscription` serves as a state indicator:
- As a `Promise`: Awaiting the followSubscription token; everything is pending.
- As a `string`: The `follow` subscription is in action.
- As `null`: The subscription has ended due to user cancellation, a drop event, or another error.

Reply:
Yup that makes sense. Lets take what you wrote here and put it as part of a readme or wiki.

Reply:
Absolutely, I'm on board with integrating these insights into the library's documentation. It's crucial for users to understand not just the "how" but the "why" behind certain behaviors.

### Documentation vs. Implementation:
While documentation will provide clarity on expected behavior and its rationale, it's important that it doesn't delve into the specificities of the implementation (like I did in the previous comment). This allows for flexibility in evolving the codebase without frequent documentation overhauls.

### Tests as Documentation:
In the absence of detailed documentation, tests can serve as a practical form of documentation. I personally approach library docs with caution, often leaning on the library's tests to decipher the actual behavior. For a low-level library like this, the documentation should ideally explain the reasoning behind the test behaviors. Meaning that I would rather ironing out the tests before we get started with the docs.

### A Practical Insight:
Taking the `followSubscription` implementation as an example: someone might wonder about the re-assignment of its resolved value. If we were to comment out [this specific line](https://github.com/paritytech/capi/blob/c6c697ba18e18138f25533d443e82d78d642e5be/packages/substrate-client/src/chainhead/chainhead.ts#L85), it might seem superficially like nothing's amiss. However, tests would fail, illustrating that the public API's behavior has changed. To pass the tests again, we'd need to artificially introduce delays with constructs like `await Promise.resolve()`. This example underscores the importance of tests in elucidating library behavior.

### Moving Forward:
Given the implicit way tests can document behaviors, it's all the more essential to ensure our documentation explicitly describes both what the library does and why it does it. Ensuring transparency and clarity in our design rationale is key.

So, yes 💯 : this behavior (and the reasoning behind it) should be explicitly documented, but "one step at a time." 😊